### PR TITLE
64 bit time_t

### PR DESCRIPTION
**Pull request by ellert**
Debian recently migrated to using a 64 bit time_t on most of their 32 bit architectures.
This leads to issues where a time_t variable is used in printf or sscanf.
This PR addresses those issues (and a few cases with size_t vs. long that I found when checking logs on 32 bit).

### CODE REVIEW

**Comment by ellert** [comment]
This took a few iterations to find all corner cases, but I think it is complete now. At least there are no more compiler warnings related to the change, and the tests succeed on all debian architectures for all packages.

File: gass/transfer/source/library/globus_gass_transfer_http.c [hunk 1]
```diff
@@ -4653,7 +4653,7 @@ globus_l_gass_transfer_http_construct_request(
 
 	    sprintf(cmd + strlen(cmd),
 		    GLOBUS_L_CONTENT_LENGTH_HEADER,
-		    length);
+		    (long) length);
```
Comment (msalle) [inline gass/transfer/source/library/globus_gass_transfer_http.c:4656]: not sure this is safe: `length` is a `globus_size_t` which is `size_t` which is unsigned plus it might be bigger than an signed long (and is actually a long long on some platforms), meaning it could overflow when cast to long: i.e. `(long)length` could be negative, meaning the number of digits in lines 4633-4637 above isn't taking into account a leading minus sign?

File: gass/transfer/source/library/globus_gass_transfer_http.c [hunk 2]
```diff
@@ -4743,7 +4743,7 @@ globus_l_gass_transfer_http_construct_request(
 		    proto->url.host);
 	    sprintf((char *) cmd + strlen(cmd),
 		    GLOBUS_L_CONTENT_LENGTH_HEADER,
-		    length);
+		    (long) length);
```
Comment (msalle) [inline gass/transfer/source/library/globus_gass_transfer_http.c:4746]: see comment above on line 4656

**Review by msalle — CHANGES_REQUESTED** [review]
Other than my two comments I think it looks all fine to me.

File: gass/transfer/source/library/globus_gass_transfer_http.c (hunk 1 above)
Comment (johnny-mnemonic) [reply gass/transfer/source/library/globus_gass_transfer_http.c:4656]: @ellert, @msalle: Sorry, totally forgot about that on the weekend.

I tried to find out what the exact definition of `size_t` is on my local machine (x86_64 based), it's defined in `stddef.h` (`/usr/lib/gcc/x86_64-linux-gnu/9/include/stddef.h` on my machine) like so:

```
[...]
#ifndef __SIZE_TYPE__
#define __SIZE_TYPE__ long unsigned int
#endif
#if !(defined (__GNUG__) && defined (size_t))
typedef __SIZE_TYPE__ size_t;
[...]
```
Using the following command (from https://unix.stackexchange.com/a/451235) confirms that:
```
$ echo '#include <stddef.h>' | gcc -xc -E - | grep size_t
typedef long unsigned int size_t;
```
And looking into for example http://ftp.de.debian.org/debian/pool/main/g/gcc-12/libgcc-12-dev_12.2.0-14_i386.deb shows that it is also defined like that on i386 (i686 actually).

So it's already a `long` data type. Would the cast to `long` then make any difference?

Comment (fscheiner) [reply gass/transfer/source/library/globus_gass_transfer_http.c:4656]: @ellert, @msalle: Sorry, totally forgot about that on the weekend.

I tried to find out what the exact definition of `size_t` is on my local machine (x86_64 based), it's defined in `stddef.h` (`/usr/lib/gcc/x86_64-linux-gnu/9/include/stddef.h` on my machine) like so:

```
[...]
#ifndef __SIZE_TYPE__
#define __SIZE_TYPE__ long unsigned int
#endif
#if !(defined (__GNUG__) && defined (size_t))
typedef __SIZE_TYPE__ size_t;
[...]
```
Using the following command (from https://unix.stackexchange.com/a/451235) confirms that:
```
$ echo '#include <stddef.h>' | gcc -xc -E - | grep size_t
typedef long unsigned int size_t;
```
And looking into for example http://ftp.de.debian.org/debian/pool/main/g/gcc-12/libgcc-12-dev_12.2.0-14_i386.deb shows that it is also defined like that on i386 (i686 actually).

So it's already a `long` data type. Would the cast to `long` then make any difference?

Comment (msalle) [reply gass/transfer/source/library/globus_gass_transfer_http.c:4656]: Hi Frank,
two points:
1) your test is linux specific, not sure if it's generic on all platforms (I'm pretty sure there are platforms it's a long long).
2) more importantly: casting an *unsigned* long to a *signed* long can swap the sign if it's bigger than half the MAX

Comment (fscheiner) [reply gass/transfer/source/library/globus_gass_transfer_http.c:4656]: But is `long unsigned int` equal to `unsigned long`?

Comment (msalle) [reply gass/transfer/source/library/globus_gass_transfer_http.c:4656]: yes, they are identical. long unsigned and unsigned long can be swapped and long int is the same as long.

Comment (fscheiner) [reply gass/transfer/source/library/globus_gass_transfer_http.c:4656]: Ok, thanks for clarification.

Comment (fscheiner) [reply gass/transfer/source/library/globus_gass_transfer_http.c:4656]: I wonder how Debian handles `size_t` in their t64 transition. Need to check a few package sources maybe...

Comment (ellert) [reply gass/transfer/source/library/globus_gass_transfer_http.c:4656]: The GLOBUS_L_CONTENT_LENGTH_HEADER used here is defined as:

https://github.com/gridcf/gct/blob/e479e5165becde3f6d7cd64e072b8845ca8ad306/gass/transfer/source/library/globus_l_gass_transfer_http.h#L131

I.e. it uses %ld, and therfore expects a signed long. Whether the value is printed as signed or unsigned depends on the format string, not the variable type.
```
$ cat print.c
#include <stdio.h>
int main() {
  long l = -1;
  printf ("%ld\n", l);
  printf ("%lu\n", l);
}
$ gcc -o print print.c 
$ ./print
-1
18446744073709551615
```
So if the size_t is cast to a (signed) long or an unsigned long doesn't matter, it will be printed as a signed long in either case, because that is what the format string says.

The macro GLOBUS_L_CONTENT_LENGTH_HEADER is only used in this file. It is used 5 times. The first 3 already uses a cast to long. This commit adds the same cast to the remaining 2.

An alternative fix would be to do it the opposite way: Redefine the macro to use %zu instead of %ld and remove the 3 existing casts.

On 64 bit Linux size_t is a long.
On 32 bit Linux size_t is an int. Which is why it gives a compiler warning since it doesn't match the %ld format string. int and long are both 32 bits here, but it still complains.

On 64 bit windows a size_t is a long long, because long on 64 bit windows is 32 bit. But I don't know if this is relevant.

Comment (ellert) [reply gass/transfer/source/library/globus_gass_transfer_http.c]: > An alternative fix would be to do it the opposite way: Redefine the macro to use %zu instead of %ld and remove the 3 existing casts.

I have implemented this.

Comment (ellert) [reply gass/transfer/source/library/globus_gass_transfer_http.c]: > ... that it is also defined like that on i386 (i686 actually).
size_t is not the same for x86_64 and ix86:
```
$ echo '#include <stddef.h>' | gcc -xc -E - | grep size_t
typedef long unsigned int size_t;
$ echo '#include <stddef.h>' | gcc -m32 -xc -E - | grep size_t
typedef unsigned int size_t;
```

Comment (ellert) [reply gass/transfer/source/library/globus_gass_transfer_http.c]: > I wonder how Debian handles `size_t` in their t64 transition. Need to check a few package sources maybe...

There are no change to size_t due to the changes in time_t.
There is potentially a change to off_t, since it is not possible to use 64 bit time_t without also using 64 bit off_t. GCT is already using 64 bit off_t (-D_FILE_OFFSET_BITS=64) on 32 bit architectures. So enabling 64 bit time_t doesn't change off_t for GCT.

Comment (fscheiner) [reply gass/transfer/source/library/globus_gass_transfer_http.c]: > > ... that it is also defined like that on i386 (i686 actually).
> > size_t is not the same for x86_64 and ix86:
> 
> ```
> $ echo '#include <stddef.h>' | gcc -xc -E - | grep size_t
> typedef long unsigned int size_t;
> $ echo '#include <stddef.h>' | gcc -m32 -xc -E - | grep size_t
> typedef unsigned int size_t;
> ```

Thanks for checking, I only looked into the definition in the compiler dev package on x86. I assume `__SIZE_TYPE__` or even `size_t` is then already defined differently on x86 by libc or kernel before the default definition (`long unsigned int`) could come into effect.

Comment (msalle) [reply gass/transfer/source/library/globus_gass_transfer_http.c]: > > An alternative fix would be to do it the opposite way: Redefine the macro to use %zu instead of %ld and remove the 3 existing casts.
> 
> I have implemented this.

thanks, that indeed seems like the right fix!

**Review by msalle — APPROVED** [review]
I agree with the new version, all looks fine to me
